Graphs/README.md

```diff
@@ -0,0 +1,188 @@
+# Dealing with Graphs
+
+**TODO**: Need to split this (and other sections) into "using recursion schemes" and "understanding recursion schemes" somehow.
+
+So, let's talk about how we want to deal with graphs ... first, we'd like to maintain as much as possible the same abstractions we use when dealing with trees. That is, algebras and pattern functors. This means that given the same pattern functor, we could build either a tree or a graph (just as we can build either a finite or infinite structure from the same pattern functor). And we'd like to be able to apply the same algebras (but maybe not get the same results -- we'll come back to that point).
```

Comment:
I think we could take inspiration from 7 Sketches using the same idea of some problem statement. So in this case, we should pick some problem (or set of problems) that can be solved using graphs. This will provide motivation as to why we introduce graphs. Maybe since we're comparing them to trees we can show how trees might not suffice for this problem, or how it can be generalised from a tree to a graph, etc.

I also think that more introduction definitions graphs, adjacency lists, and anything else would be helpful. Lot's of diagrams will be cool here because graphs are very visual creatures :smile:

Reply:
The "standard" example for graphs here is trees with bindings (like progarmming languages). On the one hand, programmers should be familiar with that notion, on the other, the problems to solve with them may be too compiler-y.

Perhaps the way we used them at Formation for offers would be a better example?

Reply:
> Perhaps the way we used them at Formation for offers would be a better example?

Probably best to avoid that in case of IP conflict or something :sweat_smile: What about doing calculations on networks where peers are connected? (Not sure what inspired this :stuck_out_tongue_closed_eyes:)

Reply:
Well, I wouldn't want it to _be_ the Formation stuff, but just something like how we encoded the logic.

I'd be interested in the network stuff, but it's important (at least for early examples) that we're talking about finite directed acyclic graphs, and I imagine peer-to-peer fails at least one or two of those (but maybe not for specific use cases).

Reply:
> Well, I wouldn't want it to be the Formation stuff, but just something like how we encoded the logic.

Ya, cool just making sure :smile: Sandy had a scavenger hunt example in his ADD book that was similar in fashion. If we wanted inspiration from him.

> we're talking about finite directed acyclic graphs

Ah right. We'd get the finite part by looking at a portion of the graph, say with a cutoff. Acyclic would not be guaranteed. But we _could_ look at something like Pijul for inspiration since changesets are DAGs :) I'm sure there are lots of examples we could list out and pick one or two :grin: